### PR TITLE
Upgrade flake8 to 6.0.0 

### PR DESCRIPTION
**Pull request by mpearce25**
Addresses: https://github.com/pytorch/vision/issues/7200 

# Changes

- Upgrade flake8 version 
- Match flake8 versions between CI and contributing guide
- Move flake8 install cmd to clearer location in contributing guide 

## Typing updates
For `torchvision/models/detection/generalized_rcnn.py` type hints were: 
* Added in https://github.com/pytorch/vision/pull/4631
* Removed in https://github.com/pytorch/vision/pull/4883 due to Phabricator issues
* Added back here for compatibility with upgraded flake8 version. Will see if Phabrictor issues have been resolved. If not I can add flake8 ignores to import lines. **If these work is it preferable to add the proper type hints back in a separate PR and then upgrade flake8?**

# Testing

- [x] flake8
- [x] mypy

### CODE REVIEW

**Comment by pmeier** [comment]
Hey @mpearce25 and thanks for the PR. We are currently on a very tight schedule for some stuff that we need to finish until end of week for the next release. I'll come back to this next week. In case I forget, feel free to ping me.

**Comment by justinchuby** [comment]
Have we considered using lintrunner and ruff in this project?

**Comment by johnnynunez** [comment]
> Hey @mpearce25 and thanks for the PR. We are currently on a very tight schedule for some stuff that we need to finish until end of week for the next release. I'll come back to this next week. In case I forget, feel free to ping me.

why not to move to ruff? https://github.com/charliermarsh/ruff

**Comment by pmeier** [comment]
> Have we considered using lintrunner and ruff in this project?

Yes, but it is not a priority right now. There is a push from Nova to unite the CI system across all domain libraries, but linting is at the bottom of their list.

> why not to move to ruff? https://github.com/charliermarsh/ruff

Because that costs time that we currently don't have. `ruff` labels itself as [alpha product](https://github.com/charliermarsh/ruff/blob/3ad257cfea23e34698511eb6f32480fc65984f7d/pyproject.toml#L26) in contrast to `flake8`, which is stable since forever. I personally use `ruff` for my own small projects, but for projects as large as PyTorch there is a lot more to consider.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -83,7 +83,7 @@ Instead of relying directly on `black` however, we rely on
 [ufmt](https://github.com/omnilib/ufmt), for compatibility reasons with Facebook
 internal infrastructure.
 
-To format your code, install `ufmt` with `pip install ufmt==1.3.2 black==22.3.0 usort==1.0.2` and use e.g.:
+To format your code, install `ufmt` with `pip install ufmt==1.3.2 black==22.3.0 usort==1.0.2 flake8=6.0.0` and use e.g.:
```
Comment (pmeier) [inline CONTRIBUTING.md:86]: Since this section is about formatting, can we not include `flake8` here, but rather have a separate installation command below where `flake8` usage is explained? Alternatively, we could also rephrase this to not start with 

> To format your code, install `ufmt`

**Review by pmeier — COMMENTED** [review]
Thanks @mpearce25 for the PR! I have a small comment inline. Otherwise LGTM if CI is green.

File: CONTRIBUTING.md (hunk 1 above)
Comment (mpearce25) [reply CONTRIBUTING.md]: I split up the formatting and linting sections. 

Also promoted the "typing" section to same level as formatting/linting and unit tests.  Mypy operates independently from fromatting/linting and I wanted to avoid confusion in the pre-commit hooks section.

lmk if these changes look good. 

CI is failing same as other PRs. Issues seem unrelated, but I'm fine waiting until everything is green again to avoid compounding any CI issues.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -49,7 +49,7 @@ python setup.py develop
 # MACOSX_DEPLOYMENT_TARGET=10.9 CC=clang CXX=clang++ python setup.py develop
 # for C++ debugging, please use DEBUG=1
 # DEBUG=1 python setup.py develop
-pip install flake8 typing mypy pytest pytest-mock scipy
+pip install typing mypy pytest pytest-mock scipy
```
Comment (pmeier) [inline CONTRIBUTING.md:52]: Since we are on Python >= 3.8 now, we can drop the `typing` backport

```suggestion
pip install mypy pytest pytest-mock scipy
```

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -97,11 +96,17 @@ files that were edited in your PR with e.g.:
 ufmt format `git diff main --name-only`
 ```
 
-Similarly, you can check for `flake8` errors with `flake8 torchvision`, although
-they should be fairly rare considering that most of the errors are automatically
-taken care of by `ufmt` already.
+#### Linting
+The codebase's pep8 compliance is checked with [flake8](https://flake8.pycqa.org/en/latest/).
```
Comment (pmeier) [inline CONTRIBUTING.md:100]:
```suggestion

The codebase's pep8 compliance is checked with [flake8](https://flake8.pycqa.org/en/latest/).
```

**Review by pmeier — COMMENTED** [review]
Two nits inline. Otherwise LGTM!

File: CONTRIBUTING.md (hunk 1 above)
Comment (pmeier) [reply CONTRIBUTING.md]: > CI is failing same as other PRs. Issues seem unrelated, but I'm fine waiting until everything is green again to avoid compounding any CI issues.

Just as a heads up, this will take some time. See #7299.

Comment (mpearce25) [reply CONTRIBUTING.md]: The previous issue that led to these type hints being removed is partially hidden behind internal meta tools, but seems to basically be https://github.com/pytorch/vision/pull/4631/files#r744161323 . 

Imo it would be odd to sacrifice typing in torchvision just to avoid adding some type ignore or assert in the 3rd party code. However, lmk if that's not the case.
